File: ASCS-Vehicle/src/team/hobbyrobot/subos/logging/Logger.java
package team.hobbyrobot.subos.logging;

import java.io.Closeable;
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;

public class Logger implements Closeable
{
	private ArrayList<PrintWriter> writers = new ArrayList<PrintWriter>();
	private ArrayList<String> localLog = null;
	private boolean includeName = true;
	
	protected String name = null;
	
	public Logger()
	{
		this(null, false);
	}
	
	public Logger(String name)
	{
		this(name, false);
	}
	
	public Logger(String name, boolean saveLogToMemory)
	{
		this.name = name;
		this.localLog = saveLogToMemory ? new ArrayList<String>() : null;
		
		Runtime.getRuntime().addShutdownHook(new Thread()
		{
			@Override
			public void run()
			{
				log("Shutting down...");
				try
				{
					close();
				}
				catch (IOException e)
				{
					e.printStackTrace();
				}
			}
		});
	}

	public String[] getLog()
	{
		return localLog.toArray(new String[localLog.size()]);
	}
	
	public void setIncludeName(boolean val)
	{
		includeName = val;
	}

	public synchronized void registerEndpoint(PrintWriter writer)
	{
		writers.add(writer);
	}

	public synchronized void unregisterEndpoint(PrintWriter writer)
	{
		writers.remove(writer);
	}

	public synchronized void log(String message)
	{
		String msg = message;
		if(includeName && name != null)
			msg = name + ": " + msg;
		
		if(localLog != null)
			localLog.add(msg);

		ArrayList<PrintWriter> badWriters = new ArrayList<PrintWriter>();
		for (PrintWriter pw : writers)
		{
			try
			{
				pw.println(msg);
				pw.flush();
			}
			catch (Exception ex)
			{
				badWriters.add(pw);
			}
		}

		if (badWriters.size() > 0)
		{
			for (PrintWriter pw : badWriters)
				writers.remove(pw);
			for (int i = 0; i < badWriters.size(); i++)
				log("Writer for logger threw an exception");
		}
	}

	public static String getExceptionInfo(Exception ex)
	{
		String msg = "Error message: " + ex.getMessage() + "; Stack trace: ";

		for (StackTraceElement trace : ex.getStackTrace())
		{
			msg += "\n\t" + trace.toString();
		}

		msg += "\n; Cause: " + ex.getCause() + "; ToString: " + ex.toString();
		return msg;
	}

	@Override
	public void close() throws IOException
	{
		for (PrintWriter pw : writers)
			pw.close();
	}
	
	public Logger createSubLogger(String name)
	{
		return new SubLogger(name, this);
	}
	
	public String getName()
	{
		return name;
	}
	
	public class SubLogger extends Logger
	{
		private Logger parent;
		SubLogger(String name, Logger parent)
		{
			super();
			this.name = name;
			this.parent = parent;
			setIncludeName(false);
		}
		
		@Override
		public synchronized void log(String message) 
		{
			super.log(message);
			parent.log(name + ": " + message);
		}
		
		public Logger getParent()
		{
			return parent;
		}
	}
}

File: AutonomousStorageControlSystem/Core/Logging/Logger.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Team.HobbyRobot.ASCS.Core.Logging
{
    public class Logger : IDisposable
    {
        private LinkedList<TextWriter> writers = new LinkedList<TextWriter>();
        public IEnumerable<string> LocalLog => localLog;
        private LinkedList<string> localLog = new LinkedList<string>();

		public string Name { get; protected set; }
        public bool IncludeName { get; set; }

        public Logger() : this(null)
		{
		}

        public Logger(string name)
        {
            Name = name;
            AppDomain.CurrentDomain.ProcessExit += (s, e) => { Log("Shutting down..."); Dispose(); };
        }

        public void RegisterEndpoint(TextWriter writer)
        {
            writers.AddLast(writer);
        }

        public void UnregisterEndpoint(TextWriter writer)
        {
            writers.Remove(writer);
        }

		public virtual void Log(string message)
		{
			string msg = message;
			if (IncludeName && Name != null)
				msg = Name + ": " + msg;

			localLog.AddLast(msg);

			LinkedList<TextWriter> badWriters = new LinkedList<TextWriter>();
			foreach(TextWriter tw in writers)
			{
				try
				{
					tw.WriteLine(msg);
					tw.Flush();
				}
				catch (Exception)
				{
					badWriters.AddLast(tw);
				}
			}

			if (badWriters.Count() > 0)
			{
				foreach(TextWriter tw in badWriters)
					writers.Remove(tw);
				for (int i = 0; i < badWriters.Count(); i++)
					Log("Writer for logger threw an exception");
			}
		}

		public Logger CreateSubLogger(String name)
		{
			return new SubLogger(name, this);
		}

		public void Dispose()
        {
            foreach (TextWriter writer in writers)
                writer.Dispose();
        }

		public class SubLogger : Logger
		{
			public Logger parent { get; }
			internal SubLogger(string name, Logger parent) : base(name)
			{
				this.parent = parent;
				IncludeName = false;
			}

            public override void Log(string message)
            {
                base.Log(message);
				parent.Log(Name + ": " + message);
			}
	}
}
}
